### DISCUSSION TITLE
Custom unrestricted usage of Abstract types into Canonical types

### DISCUSSION THREAD

**Discussion by nitindeo** [google/digitalbuildings#599] [Ideas]
Hi,
Here is a suggestion to the ontology. I have atypical AHU to be modelled which includes the following abstract types.
SSPC, CHWDPSC, CHWPVM, RTM, VOCC, MOAFC, FDSTD.
There is no canonical type available which can be used. Hence the way forward shall be to make a pull request to add this new canonical type. 
If the ontology is modified so as to allow free combination of abstract types in modelling, then there will not be any need to create multiple custom canonical types. 
This will greatly provide flexibility and reduce new canonical types being requested.
I am currently used GATEWAY/PASSTHROUGH to verify all the points to via validator. The file is as below.
Can this be made as a new flexible feature in the ontology.
The yaml file is as below.

IN-MUM-DEMO:
  type: FACILITIES/BUILDING
  guid: 839402
  id: CDM/123

# Floor
IN-MUM-DEMO-FLR1:
  type: FACILITIES/FLOOR
  id: FACILITIES/a388563d-58e3-4534-b133-19389f6650a4
  guid: 987654
  connections:
    IN-MUM-DEMO: CONTAINS

AHU-1:
  type: GATEWAYS/PASSTHROUGH
  id: 12345
  guid: 12345566
  cloud_device_id: 1234567
  translation:
    supply_fan_run_command:
      present_value: points.supply_fan_run_command.present_value
      states:
        ON: "ON"
        OFF: "OFF"
    supply_fan_run_status:
      present_value: points.supply_fan_run_status.present_value
      states:
        ON: "ON"
        OFF: "OFF"
    control_mode:
      present_value: points.control_mode.present_value
      states:
        AUTO: "AUTO"
        MANUAL: "MANUAL"
        OFF: "OFF"
    power_status:
      present_value: power_status.present_value
      states:
        PRESENT: "PRESENT"
        ABSENT: "ABSENT"
    filter_alarm:
      present_value: points.filter_alarm.present_value
      states:
        ACTIVE: "ACTIVE"
        INACTIVE: "INACTIVE"
    damper_status_1:
      present_value: points.damper_status_1.present_value
      states:
        OPEN: "OPEN"
        CLOSED: "CLOSED"
    damper_status_2:
      present_value: points.damper_status_2.present_value
      states:
        OPEN: "OPEN"
        CLOSED: "CLOSED"
    schedule_run_command:
      present_value: points.schedule_run_command.present_value
      states:
        ON: "ON"
        OFF: "OFF"
    return_air_temperature_sensor:
      present_value: points.return_air_temperature_sensor.present_value
      units:
        key: pointset.points.return_air_temperature_sensor.units
        values:
          degrees_celsius: degC
    return_air_relative_humidity_sensor:
      present_value: points.return_air_relative_humidity_sensor.present_value
      units:
        key: pointset.points.return_air_relative_humidity_sensor.units
        values:
          percent_relative_humidity: '%RH'
    zone_air_voc_concentration_sensor:
      present_value: points.zone_air_voc_concentration_sensor.present_value
      units:
        key: pointset.points.zone_air_voc_concentration_sensor.units
        values:
          parts_per_million: ppm
    supply_air_dewpoint_temperature_sensor:
      present_value: points.supply_air_dewpoint_temperature_sensor.present_value
      units:
        key: pointset.points.supply_air_dewpoint_temperature_sensor.units
        values:
          degrees_celsius: degC
    supply_air_static_pressure_sensor:
      present_value: points.supply_air_static_pressure_sensor.present_value
      units:
        key: pointset.points.supply_air_static_pressure_sensor.units
        values:
          pascals: pa
    outside_air_damper_percentage_sensor:
      present_value: points.outside_air_damper_percentage_sensor.present_value
      units:
        key: pointset.points.outside_air_damper_percentage_sensor.units
        values:
          percent: '%'
    supply_air_flowrate_sensor:
      present_value: points.supply_air_flowrate_sensor.present_value
      units:
        key: pointset.points.supply_air_flowrate_sensor.units
        values:
          cubic_feet_per_minute: cfm
    ventilation_outside_air_flowrate_sensor:
      present_value: points.ventilation_outside_air_flowrate_sensor.present_value
      units:
        key: pointset.points.ventilation_outside_air_flowrate_sensor.units
        values:
          cubic_feet_per_minute: cfm
    supply_fan_speed_frequency_sensor:
      present_value: points.supply_fan_speed_frequency_sensor.present_value
      units:
        key: pointset.points.supply_fan_speed_frequency_sensor.units
        values:
          hertz: Hz
    supply_fan_current_sensor:
      present_value: points.supply_fan_current_sensor.present_value
      units:
        key: pointset.points.supply_fan_current_sensor.units
        values:
          amperes: A
    supply_fan_speed_percentage_sensor:
      present_value: points.supply_fan_speed_percentage_sensor.present_value
      units:
        key: pointset.points.supply_fan_speed_percentage_sensor.units
        values:
          percent: '%'
    supply_fan_power_sensor:
      present_value: points.supply_fan_power_sensor.present_value
      units:
        key: pointset.points.supply_fan_power_sensor.units
        values:
          watts: W         
    energy_accumulator:
      present_value: points.energy_accumulator.present_value
      units:
        key: pointset.points.energy_accumulator.units
        values:
          kilowatt_hours: KWh
    supply_fan_speed_percentage_command:
      present_value: points.supply_fan_speed_percentage_command.present_value
      units:
        key: pointset.points.supply_fan_speed_percentage_command.units
        values:
          percent: '%'
    speed_percentage_sensor:
      present_value: points.speed_percentage_sensor.present_value
      units:
        key: pointset.points.speed_percentage_sensor.units
        values:
          percent: '%'
    chilled_water_valve_percentage_command:
      present_value: points.chilled_water_valve_percentage_command.present_value
      units:
        key: pointset.points.chilled_water_valve_percentage_command.units
        values:
          percent: '%'
    chilled_water_valve_percentage_sensor:
      present_value: points.chilled_water_valve_percentage_sensor.present_value
      units:
        key: pointset.points.chilled_water_valve_percentage_sensor.units
        values:
          percent: '%'
    zone_air_voc_concentration_setpoint:
      present_value: points.zone_air_voc_concentration_setpoint.present_value
      units:
        key: pointset.points.zone_air_voc_concentration_setpoint.units
        values:
          parts_per_million: ppm
    supply_air_dewpoint_temperature_setpoint:
      present_value: points.supply_air_dewpoint_temperature_setpoint.present_value
      units:
        key: pointset.points.supply_air_dewpoint_temperature_setpoint.units
        values:
          degrees_celsius: degC
    supply_air_static_pressure_setpoint:
      present_value: points.supply_air_static_pressure_setpoint.present_value
      units:
        key: pointset.points.supply_air_static_pressure_setpoint.units
        values:
          pascals: pa
    ventilation_outside_air_flowrate_setpoint:
      present_value: points.ventilation_outside_air_flowrate_setpoint.present_value
      units:
        key: pointset.points.ventilation_outside_air_flowrate_setpoint.units
        values:
          cubic_feet_per_minute: cfm
    outside_air_damper_percentage_command:
      present_value: points.outside_air_damper_percentage_command.present_value
      units:
        key: pointset.points.outside_air_damper_percentage_command.units
        values:
          percent: '%'

**Comment by charbull** [comment]
@mschulze17 @tasodorff PTAL

**Comment by tasodorff** [comment]
the reason behind why the entity is atypical makes a difference here: if the device operates differently than a typical device in some fundamental way, then we would want to add those abstract types which fully describe its functionality. if, however, the device just doesn't send us all of the data we would expect for an entity of the type it actually is, then we would use the correct canonical type and mark the unavailable fields as 'missing'. 

a third alternative, but to be used only for rare instances, would be to commit a type which is non-canonical. to do this, we would mark `is_canonical: False` in the type definition. we would typically not allow this, and only would consider it for devices that were truly unique in some way that couldn't be adequately described by an existing type and also didn't rise to the level of being worth modeling its exact functional representation.
